granulate_utils/linux/ns.py

```diff
@@ -0,0 +1,148 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+
+import ctypes
+import os
+from pathlib import Path
+from threading import Thread
+from typing import Callable, List, Optional, TypeVar, Union
+
+from psutil import NoSuchProcess, Process
+
+T = TypeVar("T")
+
+
+class _Sentinel:
+    pass
+
+
+_SENTINEL = _Sentinel()
+
+
+def resolve_proc_root_links(proc_root: str, ns_path: str) -> str:
+    """
+    Resolves "ns_path" which (possibly) resides in another mount namespace.
+
+    If ns_path contains absolute symlinks, it can't be accessed merely by /proc/pid/root/ns_path,
+    because the resolved absolute symlinks will "escape" the /proc/pid/root base.
+
+    To work around that, we resolve the path component by component; if any component "escapes", we
+    add the /proc/pid/root prefix once again.
+    """
+    assert ns_path[0] == "/", f"expected {ns_path!r} to be absolute"
+    parts = Path(ns_path).parts
+
+    path = proc_root
+    for part in parts[1:]:  # skip the / (or multiple /// as .parts gives them)
+        next_path = os.path.join(path, part)
+        if os.path.islink(next_path):
+            link = os.readlink(next_path)
+            if os.path.isabs(link):
+                # absolute - prefix with proc_root
+                next_path = proc_root + link
+            else:
+                # relative: just join
+                next_path = os.path.join(path, link)
+        path = next_path
+
+    return path
+
+
+def is_same_ns(pid: int, nstype: str, pid2: int = None) -> bool:
+    return (
+        os.stat(f"/proc/{pid2 if pid2 is not None else 'self'}/ns/{nstype}").st_ino
+        == os.stat(f"/proc/{pid}/ns/{nstype}").st_ino
+    )
+
+
+def run_in_ns(nstypes: List[str], callback: Callable[[], T], target_pid: int = 1) -> T:
+    """
+    Runs a callback in a new thread, switching to a set of the namespaces of a target process before
+    doing so.
+
+    Needed initially for switching mount namespaces, because we can't setns(CLONE_NEWNS) in a multithreaded
+    program (unless we unshare(CLONE_NEWNS) before). so, we start a new thread, unshare() & setns() it,
+    run our callback and then stop the thread (so we don't keep unshared threads running around).
+    For other namespace types, we use this function to execute callbacks without changing the namespaces
+    for the core threads.
+
+    By default, run stuff in init NS. You can pass 'target_pid' to run in the namespace of that process.
+    """
+
+    # make sure "mnt" is last, once we change it our /proc is gone
+    nstypes = sorted(nstypes, key=lambda ns: 1 if ns == "mnt" else 0)
+
+    ret: Union[T, _Sentinel] = _SENTINEL
+    exc: Optional[BaseException] = None
+
+    def _switch_and_run():
+        try:
+            libc = ctypes.CDLL("libc.so.6")
```

Comment:
Extract to global along with the functions so that we don't `dlopen`, `dlclose`, `dlsym` all the time.